rfcs/0042-config-option.md

```diff
@@ -0,0 +1,301 @@
+- Ability to easily query arbitrary configuration values with `nix-instantiate --eval '<nixpkgs/nixos>' -A config.services.foo.config`
+- The configuration file is well formatted with the right amount of indentation everywhere
+- Usually hardcoded defaults can now be replaced by simple assignment of the `config` option, which also allows people to override those values. See the [Defaults](#defaults) section for more details.
+
+*: The only problem it doesn't solve is the coupling to upstream defaults for required defaults, see the [Defaults](#defaults) section.
+
+## Defaults
+
+Depending on how default settings matter, we need to set them differently and for different reasons:
+- If the module needs a specific value for a setting because of how the module or NixOS works (e.g. `logger = "systemd"`, because NixOS uses `systemd` for logging), then the value should *not* use `mkDefault`. This way a user can't easily override this setting (which would break the module in some way) and will have to use `mkForce` instead to change it. This also indicates that they are leaving supported territory, and will probably have to change something else to make it work again (e.g. if they set `logger = mkForce "/var/log/foo"` they'll have to change their workflow of where to look for logs).
+- If the program needs a setting to be present in the configuration file because otherwise it would fail at runtime and demand a value, the module should set this value *with* a `mkDefault` to the default upstream value, which will then be the equivalent of a starter configuration file. This allows users to easily change the value, but also enables a smooth first use of the module without having to manually set such defaults to get it to a working state. Optimally modules should Just Work (tm) by setting their `enable` option to true.
+- If the module itself needs to know the value of a configuration setting at evaluation time in order to influence other options (e.g. opening the firewall for a services port), we may set upstream's default with a `mkDefault`, even though the program might start just fine without it. This allows the module to use the configuration setting directly without having to worry whether it is set at all at evaluation time.
+
+If the above points don't apply to a configuration setting, that is the module doesn't care about the value, the program doesn't care about the setting being present and we don't need the value at evaluation time, there should be no need to specify any default value.
+
+## Additional config options
+
+Sometimes it makes sense to have an additional NixOS option for a specific configuration setting. In general this should be discussed on a case-by-case basis to judge whether it makes sense. However keep in mind that it's always possible to add more options later on, but you can't as easily remove existing options. Instances of where it can make sense are:
```

Comment:
Like @Zimmi48, I also appreciate the availability of named options in the documentation with types and brief descriptions. It truly helps when working with a system configuration. I also appreciate the that I from the type can determine how the option will be merged if I choose to split my configuration over multiple files. This is the primary issue I have with moving everything into an "opaque" option, so I like this section quite a lot. But I also note that it is unclear whether this allowance of options re-introduces any or all of the problems outlined in the Motivation section.

In general I suspect that the [Pareto principle](https://en.wikipedia.org/wiki/Pareto_principle) applies here. Something like "20% of available program options satisfies 80% of use-cases" perhaps? 🙂

The issue, I suppose, is how do we identify the options that should be treated specially? It is not clear that "it is up to the module author to decide which options qualify for this" would be sufficient since this is, as far as I know, the policy being in use historically and lead to this RFC.

Perhaps it would be possible to find some way to really identify which options are most used? Something like [Debian Popularity Contest](https://popcon.debian.org/) but for NixOS configurations would be pretty cool and could help guide the choice in which options to introduce and which to remove. I don't know how feasible something like that would be, typically I think the option names shouldn't cause information leaks but some can contain sensitive information like host and user names…

Reply:
I think this RFC is for improved and uniform treatment of options not chosen by module authors, so maybe its existence doesn't say leaving the choice of most-important options to module authors is a bad idea

Reply:
> But I also note that it is unclear whether this allowance of options re-introduces any or all of the problems outlined in the Motivation section.

Yeah I should mention something on that, obviously adding such option ties us to upstream partially again.

> The issue, I suppose, is how do we identify the options that should be treated specially? It is not clear that "it is up to the module author to decide which options qualify for this" would be sufficient since this is, as far as I know, the policy being in use historically and lead to this RFC.

I'm not sure how else we could handle it. If you have a good and realistic idea I'd love to hear it.

> Perhaps it would be possible to find some way to really identify which options are most used? Something like Debian Popularity Contest but for NixOS configurations would be pretty cool and could help guide the choice in which options to introduce and which to remove. I don't know how feasible something like that would be, typically I think the option names shouldn't cause information leaks but some can contain sensitive information like host and user names…

Cool idea (but out of scope for this RFC)! There is [the configuration collection in the wiki](https://nixos.wiki/wiki/Configuration_Collection), but it wouldn't be trivial to parse out used options. And because there's *so* many options, and everybody only uses a very small part of it, I doubt this will be useful for all modules. The only way I see this working is if we add some analytics to `nixos-rebuild` that reports used options.

Reply:
I consider the "mentioning that additional options reintroduces some problems" resolved with https://github.com/NixOS/rfcs/pull/42/commits/e9082a92e0f97613cbdc9cb40ed0f6eb5a7ccd5e